projects/ziti-console-lib/src/lib/features/visualizer/network-visualizer/network-visualizer.component.ts

```diff
@@ -0,0 +1,1337 @@
+        }
+        if (typeof nodeObj.data.type === 'undefined') {
+            //      //
+        } else {
+            if (arr.indexOf(name) < 0 && nodeObj.data.type === typeName) {
+                arr.push(name);
+            }
+        }
+
+        const childNodes = nodeObj.children ? nodeObj.children : nodeObj._children;
+        for (let i = 0; childNodes && i < childNodes.length; i++) {
+            this.fillAutoCompleteTreeNodes(childNodes[i], arr, typeName);
+        }
+
+        return arr;
+    }
+
+  initTopoView() {
+        this.autocompleteOptions = [];
+        // const duration = 750;
+        const margin = { top: 20, right: 90, bottom: 30, left: 0 };
+        const width = 1600 - margin.left - margin.right;
+        const height = 1600 - margin.top - margin.bottom;
+        this.treetooltip = d3.select('#tooltip');
+
+        this.resourceTypeError = false;
+        this.filterText = '';
+        this.resourceType = '';
+
+        const handleZoom = (e) => {
+            this.svg.attr('transform', e.transform);
+        };
+        this.zoom = d3.zoom().on('zoom', handleZoom);
+
+        this.svg = d3
+            .select('#NetworkVisualizerSVG')
+            .attr('width', '100%')
+            .attr('height', '100%')
+            .call(this.zoom)
+            .attr('viewBox', '0 ' + (-1 * (height - margin.top - margin.bottom)) / 3 + ' ' + width + ' ' + height)
+            .append('g')
+            .attr('transform', 'translate(' + -300 + ',' + 225 + ')');
+
+        d3.select('body').on('click', function () {
+            d3.select('#topocontextmenu').style('display', 'none');
+        });
+
+        d3.selectAll('g > *').remove();
+
+        // this.treemap = d3.tree().nodeSize([140, 40]);
+        this.treemap = d3.tree().nodeSize([100, 20]);
+
+        this.networkGraph.x = height / 4;
+        this.networkGraph.y = 100;
+        this.networkGraph.children.forEach(collapse);
+        this.updateTree(this.networkGraph);
+
+        _.delay(() => {
+            this.transform = d3.zoomIdentity;
+            this.transform.x = -300;
+            this.transform.y = 225;
+            this.transform.k = 1;
+            this.svg.transition().duration(750).attr('transform', this.transform);
+        }, 600);
+
+        function collapse(d) {
+            if (d.data.rootNode === 'Yes') {
+                d.data.clickProcess = 'Not Yet';
+                d._children = null;
+            }
+            if (d.children) {
+                d._children = d.children;
+                d._children.forEach(collapse);
+                d.children = null;
+            }
+        }
+  }
+
+    updateTree(source) {
+        d3.selectAll('g > *').remove();
+        const duration = 750;
+        try {
+            this.treeData = this.treemap(this.networkGraph);
+            this.nodes = this.treeData.descendants();
+            this.links = this.treeData.descendants().slice(1);
+        } catch (e) {
+          console.log('Error:', e);
```

Comment:
I'm not opposed to a try/catch here, but we might might to be a little more specific about what is implied if we catch an error at this stage in the rendering.

Reply:
(1) This is based on our experience with visualization. We retrieve all objects (identities, routers, policies, configurations, etc.) and construct a visualizer. Any missing object info or defects in these objects can result in unexpected errors when building the D3 tree. These errors cause the page to become unresponsive, requiring the user to close the window and reopen the entire application.
Since it's difficult to analyze such issues when thousands of objects are involved, a generic try-catch block is added to ensure that the page is responsive.
  
(2) yes, Logger service is more appropriate.